cli/cmd/check.go

```diff
@@ -172,3 +186,72 @@ func runChecks(w io.Writer, hc *healthcheck.HealthChecker) bool {
 
 	return hc.RunChecks(prettyPrintResults)
 }
+
+type checkCategory struct {
+	Name   string   `json:"categoryName"`
+	Checks []*check `json:"checks"`
+}
+
+type check struct {
+	Description string      `json:"description"`
+	Hint        string      `json:"hint"`
+	Result      checkResult `json:"result"`
```

Comment:
what do you think about adding json decorators to the `healthcheck.CheckResult` struct, rather than creating this similar `check` struct? that should hopefully simplify `collectJSONOutput` a bit?

added bonus, it will pick up `result.Err` for free, as i think we should include that in the json output...

```
linkerd-version
---------------
√ can determine the latest version
‼ cli is up-to-date
    unsupported version channel: dev-abcb8bb2-siggy
    see https://linkerd.io/checks/#l5d-version-cli for hints
```

```json
  {
    "categoryName": "linkerd-version",
    "checks": [
      {
        "description": "can determine the latest version",
        "hint": "",
        "result": "success"
      },
      {
        "description": "cli is up-to-date",
        "hint": "https://linkerd.io/checks/#l5d-version-cli",
        "result": "warning"
      }
    ]
  },
```

Reply:
Hmm but how would I serialize the `"result": "success"`? I believe I need an extra struct for that anyway. I mean I can add that field to `CheckResult` but I would probably prefer to have those two structs separate. Then when you change the internal struct (CheckResult) you won't break the output for users. It's kind of API because scripts will depend on it so it should be stable at all times.

Reply:
Yeah I see your point. It's not the nicest API for me though. If I got to choose what would be the easiest to consume for me as a writer of the script consuming this, I would probably go this way. And since we already have the structs in place... my choice would be to stick to them. But it's you guys's call in the end... :) I am just expressing my opinion

Reply:
After playing with JSON output from both the user-facing `cmd.check` struct and marshaling directly from `healthcheck.CheckResult`, I agree with you that a user-facing version is preferable, as it will make consumption via scripts easier.

My primary concern with two sets of structs was maintenance overhead, specifically the risk of them diverging when introducing new fields/features. To that end, mind adding some comments above both `cmd.check` and `healthcheck.CheckResult` indicating how these two structs are related?

A few more changes:
- include the error message if a check fails
- only render `hint` and `error` when they are non-empty
- render a top-level `success` boolean, to allow for very simple scripting, rather than having to iterate over all the checks to validate whether it passed:
    ```json
    {
      "success": false,
      "categories": [
        {
          "categoryName": "kubernetes-api",
          "checks": [
    ```

Here's a diff encapsulating the above changes, let me know what you think:
```diff
diff --git a/cli/cmd/check.go b/cli/cmd/check.go
index 58ecb123..bfd61d51 100644
--- a/cli/cmd/check.go
+++ b/cli/cmd/check.go
@@ -194,14 +194,22 @@ func runChecksTable(wout io.Writer, hc *healthcheck.HealthChecker) bool {
 	return success
 }
 
+type checkOutput struct {
+	Success    bool             `json:"success"`
+	Categories []*checkCategory `json:"categories"`
+}
+
 type checkCategory struct {
 	Name   string   `json:"categoryName"`
 	Checks []*check `json:"checks"`
 }
 
+// check is a user-facing version of `healthcheck.CheckResult`, for output via
+// `linkerd check -ojson`.
 type check struct {
 	Description string      `json:"description"`
-	Hint        string      `json:"hint"`
+	Hint        string      `json:"hint,omitempty"`
+	Error       string      `json:"error,omitempty"`
 	Result      checkResult `json:"result"`
 }
 
@@ -214,19 +222,19 @@ const (
 )
 
 func runChecksJSON(wout io.Writer, werr io.Writer, hc *healthcheck.HealthChecker) bool {
-	var outputJSON []*checkCategory
+	var categories []*checkCategory
 
 	collectJSONOutput := func(result *healthcheck.CheckResult) {
 		categoryName := string(result.Category)
-		if outputJSON == nil || outputJSON[len(outputJSON)-1].Name != categoryName {
-			outputJSON = append(outputJSON, &checkCategory{
+		if categories == nil || categories[len(categories)-1].Name != categoryName {
+			categories = append(categories, &checkCategory{
 				Name:   categoryName,
 				Checks: []*check{},
 			})
 		}
 
 		if !result.Retry {
-			currentCategory := outputJSON[len(outputJSON)-1]
+			currentCategory := categories[len(categories)-1]
 			// ignore checks that are going to be retried, we want only final results
 			status := checkSuccess
 			if result.Err != nil {
@@ -236,21 +244,29 @@ func runChecksJSON(wout io.Writer, werr io.Writer, hc *healthcheck.HealthChecker
 				}
 			}
 
-			var hint string
-			if result.Err != nil && result.HintAnchor != "" {
-				hint = fmt.Sprintf("%s%s", healthcheck.HintBaseURL, result.HintAnchor)
-			}
-
 			currentCheck := &check{
 				Description: result.Description,
 				Result:      status,
-				Hint:        hint,
 			}
+			if result.Err != nil {
+				currentCheck.Error = result.Err.Error()
+
+				if result.HintAnchor != "" {
+					currentCheck.Hint = fmt.Sprintf("%s%s", healthcheck.HintBaseURL, result.HintAnchor)
+				}
+			}
+
 			currentCategory.Checks = append(currentCategory.Checks, currentCheck)
 		}
 	}
 
 	result := hc.RunChecks(collectJSONOutput)
+
+	outputJSON := checkOutput{
+		Success:    result,
+		Categories: categories,
+	}
+
 	resultJSON, err := json.MarshalIndent(outputJSON, "", "  ")
 	if err == nil {
 		fmt.Fprint(wout, string(resultJSON))
diff --git a/pkg/healthcheck/healthcheck.go b/pkg/healthcheck/healthcheck.go
index 9760732c..afad9aae 100644
--- a/pkg/healthcheck/healthcheck.go
+++ b/pkg/healthcheck/healthcheck.go
@@ -126,7 +126,9 @@ type checker struct {
 	checkRPC func(context.Context) (*healthcheckPb.SelfCheckResponse, error)
 }
 
-// CheckResult encapsulates a check's identifying information and output
+// CheckResult encapsulates a check's identifying information and output.
+// Note there exists an analogous user-facing type, `cmd.check`, for output via
+// `linkerd check -ojson`.
 type CheckResult struct {
 	Category    CategoryID
 	Description string
```